rust/kernel/types.rs

```diff
@@ -688,3 +690,168 @@ pub enum Either<L, R> {
     /// Constructs an instance of [`Either`] containing a value of type `R`.
     Right(R),
 }
+
+/// A type that represents a duration in nanoseconds.
+pub struct Nanoseconds(u64);
+
+impl Nanoseconds {
+    /// Creates a [`Nanoseconds`] data structure from an integer number
+    /// of nanoseconds.
+    pub fn from_int(nsecs: u64) -> Self {
+        Nanoseconds(nsecs)
+    }
+
+    /// Converts a [`Nanoseconds`] data structure to an integer
+    /// representation of the number of nanoseconds.
+    ///
+    /// This is primarily useful when preparing to pass an integer value over
+    /// the FFI boundary.
+    pub fn as_int(&self) -> u64 {
+        self.0
+    }
+}
+
+impl From<Duration> for Nanoseconds {
+    // Any duration that can't be represented by a `u64` nanoseconds representation
+    // will be rounded down to `u64::MAX`.
+    fn from(duration: Duration) -> Self {
+        let nanos = duration.as_nanos();
+        // If `nanos` is chosen, `nanos` is less than or equal to `u64::MAX` which
+        // means that the conversion is valid and lossless.
+        let min_nanos = min(nanos, u128::from(u64::MAX)) as u64;
+        Nanoseconds::from_int(min_nanos)
+    }
+}
+
+impl From<Nanoseconds> for Duration {
+    fn from(duration: Nanoseconds) -> Self {
+        Duration::from_nanos(duration.as_int())
+    }
+}
+
+/// A type that represents a duration in microseconds.
+pub struct Microseconds(u64);
+
+impl Microseconds {
+    /// Creates a [`Microseconds`] data structure from an integer number
+    /// of microseconds.
```

Comment:
I would avoid calling this a data structure, even if technically it is... :)

A bit closer to what `Duration` writes is:

```suggestion
    /// Creates a new [`Microseconds`] from an integer number
    /// of microseconds.
```